Data/ByteString/Lazy/ReadInt.hs

```diff
@@ -0,0 +1,373 @@
+{-# INLINE readWord8 #-}
+readWord8 = _readWord @Word8
+
+-- | Internal polymorphic wrapper
+_readWord :: forall a b. (Integral a, Bounded a, a ~ I2W_ b)
+          => ByteString -> Maybe (a, ByteString)
+_readWord bs = case readWord64 bs of
+    Just (w64, str)
+        | w64 <= wmax64 @a
+       -> let !w = fromIntegral @Word64 @a w64
+           in Just (w, str)
+    _  -> Nothing
+{-# SPECIALISE _readWord :: ByteString -> Maybe (Word, ByteString) #-}
+{-# SPECIALISE _readWord :: ByteString -> Maybe (Word8, ByteString) #-}
+{-# SPECIALISE _readWord :: ByteString -> Maybe (Word16, ByteString) #-}
+{-# SPECIALISE _readWord :: ByteString -> Maybe (Word32, ByteString) #-}
+
+----- Arbitrary precision functions
+
+-- | readNatural reads a 'Natural' number from the beginning of the
+-- 'ByteString'.  If there is no natural number at the beginning of the string,
+-- it returns 'Nothing', otherwise it just returns the number read, and the
+-- rest of the string.
+--
+-- 'readNatural' does not ignore leading whitespace, the value must start with
+-- a decimal digit immediately at the beginning of the input string.  Leading
+-- @+@ signs are not accepted.
+--
+-- ==== __Examples__
+-- >>> readNatural "000111222333444555666777888999 all done"
+-- Just (111222333444555666777888999," all done")
+-- >>> readNatural "+000111222333444555666777888999 explicit sign"
+-- Nothing
+-- >>> readNatural "not a decimal number")
+-- Nothing
+--
+readNatural :: ByteString -> Maybe (Natural, ByteString)
+#ifdef BYTESTRING_STRICT
+readNatural (BI.BS fp len) = BI.accursedUnutterablePerformIO $
+    BI.unsafeWithForeignPtr fp $ \ptr -> do
+        let end = ptr `plusPtr` len
+        (!n, !a) <- alldigits end ptr 0 0
+        if n < len
+            then let rest = BI.BS (fp `BI.plusForeignPtr` n) (len - n)
+                  in return $! result n a rest
+            else return $! result n a empty
+#else
+readNatural = loop 0 0
+  where
+    loop !nbytes !acc = \ str -> case str of
+        Empty -> result nbytes acc str
+        Chunk c cs -> case B.length c of
+            0 -> loop nbytes acc cs -- skip empty segment
+            l -> case doChunk acc c of
+                 (0, _) -- no more digits
+                     -> result nbytes acc str
+                 (n, a)
+                     | n < l -- input not entirely digits
+                       -> result (nbytes + n) a $ Chunk (B.drop n c) cs
+                     | otherwise
+                       -- read more digits from the remaining chunks
+                       -> loop (nbytes + n) a cs
+
+    doChunk acc (BI.BS fp len) =
+        BI.accursedUnutterablePerformIO $
+            BI.unsafeWithForeignPtr fp $ \ptr ->
+                alldigits (ptr `plusPtr` len) ptr 0 acc
+#endif
+
+-- | readInteger reads an Integer from the beginning of the ByteString.  If
+-- there is no integer at the beginning of the string, it returns Nothing,
+-- otherwise it just returns the integer read, and the rest of the string.
```

Comment:
```suggestion
-- | 'readInteger' reads an 'Integer' from the beginning of the 'ByteString'.  If
-- there is no integer at the beginning of the string, it returns 'Nothing',
-- otherwise it just returns the integer read, and the rest of the string.
```